File: sdk/backd-js/tests/errors.test.ts
import { describe, it, expect } from 'vitest'
import {
  BackdError,
  AuthError,
  QueryError,
  FunctionError,
  StorageError,
  NetworkError,
  parseError,
} from '../src/errors.ts'

describe('parseError', () => {
  it('maps UNAUTHORIZED to AuthError', () => {
    const err = parseError('UNAUTHORIZED', 'not authorized', 401)
    expect(err).toBeInstanceOf(AuthError)
    expect(err.code).toBe('UNAUTHORIZED')
    expect(err.status).toBe(401)
  })

  it('maps SESSION_EXPIRED to AuthError', () => {
    const err = parseError('SESSION_EXPIRED', 'expired', 401)
    expect(err).toBeInstanceOf(AuthError)
    expect(err.code).toBe('SESSION_EXPIRED')
  })

  it('maps TOO_MANY_REQUESTS to AuthError', () => {
    const err = parseError('TOO_MANY_REQUESTS', 'slow down', 429)
    expect(err).toBeInstanceOf(AuthError)
  })

  it('maps FUNCTION_NOT_FOUND to FunctionError', () => {
    const err = parseError('FUNCTION_NOT_FOUND', 'not found', 404)
    expect(err).toBeInstanceOf(FunctionError)
  })

  it('maps FUNCTION_TIMEOUT to FunctionError', () => {
    const err = parseError('FUNCTION_TIMEOUT', 'timeout', 504)
    expect(err).toBeInstanceOf(FunctionError)
  })

  it('maps FUNCTION_ERROR to FunctionError', () => {
    const err = parseError('FUNCTION_ERROR', 'failed', 500)
    expect(err).toBeInstanceOf(FunctionError)
  })

  it('maps STORAGE_DISABLED to StorageError', () => {
    const err = parseError('STORAGE_DISABLED', 'no storage', 501)
    expect(err).toBeInstanceOf(StorageError)
  })

  it('maps FORBIDDEN (403) to QueryError', () => {
    const err = parseError('FORBIDDEN', 'denied', 403)
    expect(err).toBeInstanceOf(QueryError)
  })

  it('maps NOT_FOUND (404) to QueryError', () => {
    const err = parseError('NOT_FOUND', 'missing', 404)
    expect(err).toBeInstanceOf(QueryError)
  })

  it('maps VALIDATION_ERROR (422) to QueryError', () => {
    const err = parseError('VALIDATION_ERROR', 'invalid', 422)
    expect(err).toBeInstanceOf(QueryError)
  })

  it('maps unknown codes to BackdError', () => {
    const err = parseError('INTERNAL_ERROR', 'oops', 500)
    expect(err).toBeInstanceOf(BackdError)
    expect(err).not.toBeInstanceOf(AuthError)
    expect(err).not.toBeInstanceOf(QueryError)
  })
})

describe('NetworkError', () => {
  it('has code NETWORK_ERROR and status 0', () => {
    const err = new NetworkError('connection failed')
    expect(err.code).toBe('NETWORK_ERROR')
    expect(err.status).toBe(0)
    expect(err.detail).toBe('connection failed')
  })
})

describe('error hierarchy', () => {
  it('AuthError extends BackdError', () => {
    const err = new AuthError('UNAUTHORIZED', 'test', 401)
    expect(err).toBeInstanceOf(BackdError)
    expect(err).toBeInstanceOf(Error)
  })

  it('QueryError has table property', () => {
    const err = new QueryError('NOT_FOUND', 'missing', 404, 'orders')
    expect(err.table).toBe('orders')
  })

  it('FunctionError has fn property', () => {
    const err = new FunctionError('FUNCTION_ERROR', 'failed', 500, 'send_invoice')
    expect(err.fn).toBe('send_invoice')
  })
})

File: sdk/backd-go/errors_test.go
package backd

import (
	"errors"
	"testing"
)

func TestParseError_AuthCodes(t *testing.T) {
	for _, code := range []string{"UNAUTHORIZED", "SESSION_EXPIRED", "TOO_MANY_REQUESTS"} {
		t.Run(code, func(t *testing.T) {
			err := parseError(code, "test", 401)
			var authErr *AuthError
			if !errors.As(err, &authErr) {
				t.Errorf("expected AuthError for code %s, got %T", code, err)
			}
			if authErr.Code != code {
				t.Errorf("expected code %s, got %s", code, authErr.Code)
			}
		})
	}
}

func TestParseError_FunctionCodes(t *testing.T) {
	for _, code := range []string{"FUNCTION_NOT_FOUND", "FUNCTION_TIMEOUT", "FUNCTION_ERROR"} {
		t.Run(code, func(t *testing.T) {
			err := parseError(code, "test", 500)
			var fnErr *FunctionError
			if !errors.As(err, &fnErr) {
				t.Errorf("expected FunctionError for code %s, got %T", code, err)
			}
		})
	}
}

func TestParseError_QueryCodes(t *testing.T) {
	tests := []struct {
		code   string
		status int
	}{
		{"FORBIDDEN", 403},
		{"NOT_FOUND", 404},
		{"VALIDATION_ERROR", 422},
	}
	for _, tt := range tests {
		t.Run(tt.code, func(t *testing.T) {
			err := parseError(tt.code, "test", tt.status)
			var qErr *QueryError
			if !errors.As(err, &qErr) {
				t.Errorf("expected QueryError for status %d, got %T", tt.status, err)
			}
		})
	}
}

func TestParseError_UnknownCode(t *testing.T) {
	err := parseError("INTERNAL_ERROR", "something broke", 500)
	var backdErr *BackdError
	if !errors.As(err, &backdErr) {
		t.Fatalf("expected BackdError, got %T", err)
	}
	var authErr *AuthError
	if errors.As(err, &authErr) {
		t.Error("should not be AuthError")
	}
}

func TestBackdError_Error(t *testing.T) {
	err := &BackdError{Code: "TEST", Detail: "detail", Status: 400}
	want := "TEST: detail"
	if err.Error() != want {
		t.Errorf("got %q, want %q", err.Error(), want)
	}
}

func TestErrorHierarchy(t *testing.T) {
	authErr := &AuthError{BackdError: BackdError{Code: "UNAUTHORIZED", Detail: "test", Status: 401}}
	var backdErr *BackdError
	if !errors.As(authErr, &backdErr) {
		t.Error("AuthError should be unwrappable to BackdError")
	}

	queryErr := &QueryError{BackdError: BackdError{Code: "NOT_FOUND", Detail: "missing", Status: 404}, Table: "orders"}
	if queryErr.Table != "orders" {
		t.Errorf("expected table orders, got %s", queryErr.Table)
	}

	fnErr := &FunctionError{BackdError: BackdError{Code: "FUNCTION_ERROR", Detail: "fail", Status: 500}, Fn: "send"}
	if fnErr.Fn != "send" {
		t.Errorf("expected fn send, got %s", fnErr.Fn)
	}

	netErr := &NetworkError{BackdError: BackdError{Code: "NETWORK_ERROR", Detail: "timeout", Status: 0}}
	if netErr.Status != 0 {
		t.Errorf("expected status 0, got %d", netErr.Status)
	}
}
